### ISSUE TITLE
Cilium without kube-proxy and systemd 247 is broken on Debian Bullseye

### ISSUE THREAD

**Issue by uykusuz** [cilium/cilium#14658]
## Bug report

Installing cilium on a fresh kubernetes installation without kube-proxy on debian bullseye with systemd 247 doesn't work: coredns won't come up because the API server is not reachable. It keeps saying "still waiting on kubernetes". Everything comes up except coredns.

It's related to kube-proxy: it works with kube-proxy enabled.

**General Information**

- Linux: debian 10 with bullseye repos enabled
- Cilium version (run `cilium version`): 1.9.1
- Kernel version (run `uname -a`): 5.10.0-1-amd64
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...): kubectl: 1.20.2
- Link to relevant artifacts (policies, deployments scripts, ...)
- Generate and upload a system zip: 
[cilium-sysdump-20210119-220818.zip](https://github.com/cilium/cilium/files/5838610/cilium-sysdump-20210119-220818.zip)

**How to reproduce the issue**

1. obtain a VM with debian bullseye
2. update the whole system (`apt-get update` and `apt-get dist-upgrade`)
2. install kubeadm
3. follow the [cilium guide](https://docs.cilium.io/en/v1.9/gettingstarted/kubeproxy-free/#kubeproxy-free)
4. untaint master `kubectl taint nodes "${master_node}" node-role.kubernetes.io/master:NoSchedule-`

**More Details**

We updated our system, with that systemd from 246 to 247. We were able to update the whole system with systemd on hold on 246 and it worked. Updating then systemd alone breaks it.

"Breaks it": a service requested the API server, uses the IP directly of the server and runs into a timeout repeatedly. I suspect that's also what coredns is meaning when it says "still waiting for kubernetes".

The logs in the cilium operator or in the agent don't say anything suspicious.

[10645](https://github.com/cilium/cilium/issues/10645) is not helping. `rp_filter` is 0 everywhere (i.e. when doing `sysctl -a | grep \\.rp_filter`) and the mentioned option `--enable-endpoint-routes` is not helping either. Plus, we ran with systemd 246 before without issues.

**Comment by izaakschroeder** [comment]
This may be related @uykusuz: https://github.com/cilium/cilium/issues/14459 😄

**Comment by brb** [comment]
@uykusuz Thanks for the issue. Did you check the systemd commit history between 246 and 247? Also, have you tried diff'ing `sysctl net` from the both versions?

**Comment by uykusuz** [comment]
@brb I didn't diff unfortunately. Und also I don't find a way to install systemd 246 specifically. Only 241 (from the buster repos) or 247 (from the bullseye repos).

So the only thing I can give you are some outputs from our live system, where the problem is not isolated. However, I hope this helps.

Here are two outputs from two of our nodes. [sysctl-net-systemd-246.6-2.txt](https://github.com/cilium/cilium/files/5841296/sysctl-net-systemd-246.6-2.txt) and 
[sysctl-net-systemd-247.2-4.txt](https://github.com/cilium/cilium/files/5841328/sysctl-net-systemd-247.2-4.txt).

The services on the 247 node are reachable from the rest of the cluster. But the pods on the 247 node can't communicate to the outside or other nodes, pods, host IPs.

Both have cilium 1.9.1 and both have Kernel 5.10.0-1-amd64 #1 SMP Debian 5.10.4-1 (2020-12-31) x86_64 GNU/Linux.
But the 247 node uses a VLAN and thus the default network device is eno1.4000.


cilium-status on the 247 node with 
```
systemd 247 (247.2-4)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified
```

```
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.20 (v1.20.2) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Strict   [eno1, eno1.4000 (Direct Routing)]
Cilium:                 Ok       OK
NodeMonitor:            Listening for events on 16 CPUs with 64x4096 of shared memory
Cilium health daemon:   Ok   
IPAM:                   IPv4: 5/255 allocated from 10.217.4.0/24, 
BandwidthManager:       Disabled
Host Routing:           Legacy
Masquerading:           BPF   [eno1, eno1.4000]   10.217.0.0/16
Controller Status:      29/29 healthy
Proxy Status:           OK, ip 10.217.4.197, 0 redirects active on ports 10000-20000
Hubble:                 Ok              Current/Max Flows: 4096/4096 (100.00%), Flows/s: 4.90   Metrics: Disabled
Cluster health:         5/5 reachable   (2021-01-14T13:05:40Z)
```

cilium-status on the 246 node with
```
systemd 246 (246.6-2)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
```

```
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.20 (v1.20.2) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Strict   [eth0, ens10 (Direct Routing)]
Cilium:                 Ok       OK
NodeMonitor:            Listening for events on 8 CPUs with 64x4096 of shared memory
Cilium health daemon:   Ok   
IPAM:                   IPv4: 12/255 allocated from 10.217.1.0/24, 
BandwidthManager:       Disabled
Host Routing:           Legacy
Masquerading:           BPF   [eth0, ens10]   10.217.0.0/16
Controller Status:      57/57 healthy
Proxy Status:           OK, ip 10.217.1.179, 0 redirects active on ports 10000-20000
Hubble:                 Ok              Current/Max Flows: 4096/4096 (100.00%), Flows/s: 18.45   Metrics: Disabled
Cluster health:         5/5 reachable   (2021-01-14T13:07:09Z)
```

**Comment by brb** [comment]
@uykusuz I see that you have different devices:

- systemd 246: eth0, ens10
- systemd 247: eno1, eno1.4000

In the case of 246, is ens10 also a vlan dev?

**Comment by uykusuz** [comment]
@brb No, only eno1 is a vlan device.

-------- Original-Nachricht --------

**Comment by brb** [comment]
Then you might be hitting https://github.com/cilium/cilium/issues/14579.

**Comment by uykusuz** [comment]
Thank you for pointing this out! Unfortunately I think our issue is different. There are no NodePort services involved on that node. Also the node with systemd 247 above itself cannot reach others. So egress is not working correctly.

Just to make sure: besides that issue with the live system I tested this issue with a fresh installation of debian 10 (specifically creating a single node cluster in hetzner cloud based on their debian 10 image) where I could reproduce the problems we're experiencing.

So I think unfortunately it's a more general problem with systemd 247.

-------- Original-Nachricht --------

**Comment by pchaigno** [comment]
> Just to make sure: besides that issue with the live system I tested this issue with a fresh installation of debian 10 (specifically creating a single node cluster in hetzner cloud based on their debian 10 image) where I could reproduce the problems we're experiencing.

Unless that's what you did here, could you test on a system with systemd 247 but without VLAN? Despite what the title might suggest, I suspect #14579 isn't specific to NodePort services.

**Comment by uykusuz** [comment]
@pchaigno that's what I did in the bug report above: create a cloud instance in hetzner cloud with debian 10, add bullseye repos, apt-get update and apt-get dist-upgrade and install a cluster with cilium using kubeadm as described in the documentation for both tools and remove the taint for the master.

**Comment by uykusuz** [comment]
We could further pin it down to the following versions: It works on 247.1-3, but doesn't work on 247.2-4

**Comment by brb** [comment]
@uykusuz Could you check the systemd changelog between those versions?

**Comment by uykusuz** [comment]
I'm just a user of cilium, so I can't figure out important changes. But here's the changelog for the above versions:

```
systemd (247.2-4) unstable; urgency=medium

  * Move libraries and NSS modules from /lib to /usr/lib.
    Keep libpam_systemd.so installed in /lib, as PAM doesn't support
    loading PAM modules from /usr/lib on a split-usr system.

 -- Michael Biebl <biebl@debian.org>  Sat, 02 Jan 2021 17:06:01 +0100

systemd (247.2-3) unstable; urgency=medium

  * test: use modern qemu numa arguments
    This fixes TEST-36-NUMAPOLICY from the upstream autopkgtest when using
    qemu >= 5.2.
  * Increase timeout when running unit tests.
    The default timeout for individual unit tests is 30s. On slower
    architectures and environments like reproducible-builds, this sometimes
    is not enough.
    Instead of fine-tuning the timeout for each individual test, we instead
    increase the timeouts by a factor of 10.
  * init-functions, getty-static.service: Don't hard-code path to systemctl
    binary.
    This should simplify an eventual move of systemctl from /bin to
    /usr/bin.
  * getty-static.service: Skip if dbus-broker is installed.
    Check for the binary in /usr/bin as this makes it easier to move the
    service files from /lib/systemd to /usr/lib/systemd.

 -- Michael Biebl <biebl@debian.org>  Sat, 26 Dec 2020 13:33:41 +0100

systemd (247.2-2) unstable; urgency=medium

  * Switch to "unified" cgroup hierarchy (i.e. cgroupv2) (Closes: #943981)

 -- Michael Biebl <biebl@debian.org>  Mon, 21 Dec 2020 20:06:49 +0100

systemd (247.2-1) unstable; urgency=medium

  [ Balint Reczey ]
  * debian/systemd.NEWS: Mention udev rules changes in 247

  [ Michael Biebl ]
  * New upstream version 247.2
  * Rebase patches
  * test: fix fd_is_mount_point() check

 -- Michael Biebl <biebl@debian.org>  Sun, 20 Dec 2020 20:44:31 +0100

systemd (247.1-4) unstable; urgency=medium

  [ наб ]
  * debian/extra/kernel-install.d/85-initrd.install: Don't install initrd when
    an explicit path was passed (Closes: #970213)
  * debian/extra/kernel-install.d/85-initrd.install: Match initrd installation
    messages and uninstallation to 90-loaderentry.install

  [ Michael Biebl ]
  * sd-device: keep escaped strings in DEVLINK= property (Closes: #976699)

 -- Michael Biebl <biebl@debian.org>  Fri, 11 Dec 2020 20:25:31 +0100
```

**Comment by lyveng** [comment]
I'm using systemd version 247.3-1~bpo10+1 on debian buster with backports kernel installed.

```
# uname -v
#1 SMP Debian 5.10.13-1~bpo10+1 (2021-02-11)
```

```
# lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 10 (buster)
Release:	10
Codename:	buster
```

```
# systemd --version
systemd 247 (247.3-1~bpo10+1)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified
```

```
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.18 (v1.18.4) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Strict   [enp1s0 (Direct Routing)]
Cilium:                 Ok       OK
NodeMonitor:            Disabled
Cilium health daemon:   Ok   
IPAM:                   IPv4: 3/9 allocated, 
ClusterMesh:            0/0 clusters ready, 0 global-services
BandwidthManager:       EDT with BPF   [enp1s0]
Host Routing:           Legacy
Masquerading:           BPF   [enp1s0]   0.0.0.0/0
Controller Status:      24/24 healthy
Proxy Status:           OK, ip 10.40.56.114, 0 redirects active on ports 10000-20000
Hubble:                 Disabled
Cluster health:         28/28 reachable   (2021-03-04T10:39:49Z)
```

Seems to work fine for me. There is no vlan involved though.

**Comment by uykusuz** [comment]
@lyveng I used bullseye, not buster. Also no vlan involved. I just bootstrapped again a single-node cluster with cilium 1.9.4 and systemd 247.3.1, doesn't work.

**Comment by brb** [comment]
@uykusuz Did you check that it's not the https://github.com/cilium/cilium/issues/10645 issue?

**Comment by uykusuz** [comment]
This issue still persists. And it's likely to cause trouble as soon as debian bullseye goes stable. Could we please keep it open?

**Comment by uykusuz** [comment]
Good news, we can perhaps close this issue. We did more tests. On our current setup we have both containerd and docker installed. We don't tell the kubelet which to use, so (acc. to the docs), it's using docker. 

When we install containerd only, and not docker, the issue is gone. So it's either related to docker or there's some other strange relationship between docker and containerd, when both are installed.

Since docker is removed with k8s 1.22 it probably doesn't make sense for you guys to invest any more time in this?

**Comment by brb** [comment]
@uykusuz Thanks for the update! Just before we close, could you check whether you are still running the same systemd version as you reported before?

**Comment by uykusuz** [comment]
@brb No, a more recent one. I retested this issue always by bootstrapping a new node, updating everything to latest (debian bullseye) and install k8s + cilium in it.

**Comment by brb** [comment]
I'm wondering whether the systemd update actually resolved it. Could you test it with Docker + containerd after updating everything?

**Comment by uykusuz** [comment]
I see, no it didn't. I create a node, update everything, install k8s with docker and containerd installed (but w/o telling kubeadm/kubelet about it, so kubeadm/kubelet should just take docker), install cilium, coredns won't come up. 

Now in the above scenario only installing containerd and not docker, everything works. Also tested it with Ubuntu (we thought it's perhaps a compile time flag in debian's systemd build), but same case there.

So I think it's an issue with docker as CRI.

**Comment by brb** [comment]
> Also tested it with Ubuntu (we thought it's perhaps a compile time flag in debian's systemd build), but same case there.

Do you mean that it doesn't work on Ubuntu too? If yes, which Ubuntu vsn?

**Comment by uykusuz** [comment]
I'm sorry, no, I meant it also works on Ubuntu with only containerd, but also systemd 247.

**Comment by brb** [comment]
One more clarification needed. With Docker + containerd on Ubuntu with systemd 247 doesn't work, right?

**Comment by uykusuz** [comment]
I don't know, didn't test it.

**Comment by Woidla27** [comment]
Has anyone found a workaround?
Tried `net.ipv4.conf.lxc*.rp_filter = 0` without success

`Linux kube01 5.10.0-11-amd64 #1 SMP Debian 5.10.92-1 (2022-01-18) x86_64 GNU/Linux`
`systemd 247 (247.3-6)`

**Comment by DirkTheDaring** [comment]
```
net.ipv4.conf.default.rp_filter=0
net.ipv4.conf.*.rp_filter=0
```

**Comment by reneluria** [comment]
This is because docker, with cgroupsv2 uses cgroupns-mode as private by default.

To fix this issue, change this in `/etc/docker/daemon.json`

```
{
  "default-cgroupns-mode": "host"
}
```

To verify, `docker inspect` on a container should show :

```
~# docker inspect <container_id> | grep Cgroupns
            "CgroupnsMode": "host",
```
